zircon.core/common/src/main/kotlin/org/hexworks/zircon/internal/component/impl/DefaultNumberInput.kt

```diff
@@ -0,0 +1,259 @@
+import org.hexworks.zircon.api.data.Position
+import org.hexworks.zircon.api.extensions.whenEnabled
+import org.hexworks.zircon.api.extensions.whenEnabledRespondWith
+import org.hexworks.zircon.api.uievent.KeyCode
+import org.hexworks.zircon.api.uievent.KeyboardEvent
+import org.hexworks.zircon.api.uievent.KeyboardEventType
+import org.hexworks.zircon.api.uievent.MouseEvent
+import org.hexworks.zircon.api.uievent.Pass
+import org.hexworks.zircon.api.uievent.Processed
+import org.hexworks.zircon.api.uievent.UIEventPhase
+import org.hexworks.zircon.api.util.TextUtils
+import org.hexworks.zircon.internal.Zircon
+import org.hexworks.zircon.internal.behavior.impl.DefaultScrollable
+import org.hexworks.zircon.internal.component.impl.textedit.EditableTextBuffer
+import org.hexworks.zircon.internal.component.impl.textedit.cursor.MovementDirection.LEFT
+import org.hexworks.zircon.internal.component.impl.textedit.cursor.MovementDirection.RIGHT
+import org.hexworks.zircon.internal.component.impl.textedit.transformation.DeleteCharacter
+import org.hexworks.zircon.internal.component.impl.textedit.transformation.DeleteCharacter.DeleteKind.BACKSPACE
+import org.hexworks.zircon.internal.component.impl.textedit.transformation.DeleteCharacter.DeleteKind.DEL
+import org.hexworks.zircon.internal.component.impl.textedit.transformation.InsertCharacter
+import org.hexworks.zircon.internal.component.impl.textedit.transformation.MoveCursor
+import org.hexworks.zircon.internal.event.ZirconEvent
+import org.hexworks.zircon.internal.event.ZirconScope
+import kotlin.math.min
+
+abstract class DefaultNumberInput(
+        initialValue: Int,
+        private val maxValue: Int,
+        componentMetadata: ComponentMetadata,
+        protected val renderingStrategy: ComponentRenderingStrategy<NumberInput>)
+    : NumberInput,
+        Disablable by Disablable.create(),
+        DefaultComponent(
+                componentMetadata = componentMetadata,
+                renderer = renderingStrategy) {
+
+    final override var text: String
+        get() = textBuffer.getText()
+        set(value) {
+            if (value == "") {
+                textBuffer = EditableTextBuffer.create("")
+            } else if (value.length <= maxNumberLength) {
+                val clean = value.replace(Regex("[^\\d]"), "")
+                if (clean.toInt() <= maxValue) {
+                    textBuffer = EditableTextBuffer.create(clean, textBuffer.cursor)
+                    computeDigitalValue()
+                    render()
+                }
+            }
+        }
+
+    protected var textBuffer = EditableTextBuffer.create("$initialValue")
+    abstract var maxNumberLength: Int
+    private var textBeforeModifications = ""
+
+    final override val currentValueProperty = createPropertyFrom(initialValue)
+    override var currentValue: Int by currentValueProperty.asDelegate()
+
+    init {
+        this.text = "$initialValue"
+        computeDigitalValue()
```

Comment:
I think `digital` has a different meaning, maybe `computeNumberValue()` would be more expressive.